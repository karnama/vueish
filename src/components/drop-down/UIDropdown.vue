<template>
    <div ref="uiContextMenu"
         aria-haspopup="true"
         class="w-max relative">
        <slot />
        <div v-if="isOpen"
             v-bind="$attrs"
             ref="menu"
             class="menu absolute flex flex-col items-stretch shadow-lg bg-white rounded z-50 overflow-scroll"
             :style="menuStyle"
             role="group"
             @click.stop>
            <slot name="menu" />
        </div>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent, nextTick, onMounted, onUpdated, ref, watch } from 'vue';
import type { PropType } from 'vue';
import { removeClickAway, addClickAway, createClickAwayListener } from '@composables/interactivity/clickAway';
import { clickAway } from '@composables/interactivity';
import { getPxValue } from '@composables/style';

export default defineComponent({
    name: 'UIDropdown',

    directives: { clickAway },

    inheritAttrs: false,

    props: {
        /**
         * The horizontal position of the menu.
         */
        horizontal: {
            type: String as PropType<'right' | 'left'>,
            default: 'right',
            validator: (val: 'right' | 'left') => ['right', 'left'].includes(val)
        },

        /**
         * The vertical position of the menu.
         */
        vertical: {
            type: String as PropType<'top' | 'bottom'>,
            default: 'bottom',
            validator: (val: 'top' | 'bottom') => ['top', 'bottom'].includes(val)
        },

        /**
         * The max height of the menu with px or vh notation.
         */
        maxHeight: {
            type: String,
            default: '35vh'
        },

        /**
         * The width of the menu with px or vw notation.
         */
        width: {
            type: String,
            default: '250px'
        },

        /**
         * Whether the position should be where
         * the mouse is currently at or not.
         */
        atMousePosition: {
            type: Boolean,
            default: false
        }
    },

    setup(props) {
        const isOpen = ref(false);
        const uiContextMenu = ref<HTMLDivElement>();
        const menu = ref<HTMLDivElement>();
        const mousePos = ref<{ x: number; y: number } | undefined>();

        const menuStyle = computed(() => {
            const style: Partial<CSSStyleDeclaration> = {};
            if (!uiContextMenu.value || !isOpen.value) {
                return style;
            }

            const contentParameters: DOMRect = uiContextMenu.value.getBoundingClientRect();
            const menuParameters = menu.value?.getBoundingClientRect();

            style.maxHeight = props.maxHeight;
            style.width = props.width;

            if (props.atMousePosition && mousePos.value) {
                // convert viewport values to px
                const width = getPxValue(props.width);
                const maxHeight = getPxValue(props.width);

                // doesn't fit on the right
                if (document.documentElement.clientWidth - (contentParameters.x + mousePos.value.x) <= width) {
                    style.left = `${mousePos.value.x - width}px`;
                } else {
                    style.left = `${mousePos.value.x}px`;
                }

                // doesn't fit on the bottom
                if (document.documentElement.clientHeight - (contentParameters.y + mousePos.value.y) <= maxHeight) {
                    style.top = `${mousePos.value.y - (menuParameters ? menuParameters.height : maxHeight)}px`;
                } else {
                    style.top = `${mousePos.value.y}px`;
                }

                return style;
            }

            if (props.horizontal === 'left') {
                style.left = '0px';
            } else {
                style.right = '0px';
            }

            if (props.vertical === 'top') {
                style.bottom = `${contentParameters.height}px`;
            } else {
                style.top = `${contentParameters.height}px`;
            }

            return style;
        });

        const toggle = (opened?: boolean) => {
            const bool = typeof opened === 'boolean' ? opened : !isOpen.value;

            if (bool === isOpen.value) return;

            if (!bool) {
                mousePos.value = undefined;
            }

            isOpen.value = bool;
        };
        const contextMenuClick = (event: MouseEvent) => {
            event.preventDefault();
            mousePos.value = { x: event.offsetX, y: event.offsetY };
            toggle(true);
        };

        onMounted(() => {
            if (props.atMousePosition) {
                uiContextMenu.value!.addEventListener('contextmenu', contextMenuClick);
            }

            createClickAwayListener(menu.value, { value: () => toggle(false), modifiers: {} });
        });
        onUpdated(() => {
            if (props.atMousePosition) {
                uiContextMenu.value!.removeEventListener('contextmenu', contextMenuClick);
                uiContextMenu.value!.addEventListener('contextmenu', contextMenuClick);
            } else {
                uiContextMenu.value!.removeEventListener('contextmenu', contextMenuClick);
            }
        });

        watch(() => isOpen.value, async val => {
            if (val) {
                addClickAway();
                await nextTick();
                menu.value!.addEventListener('contextmenu', e => e.stopPropagation());
            } else {
                menu.value!.removeEventListener('contextmenu', e => e.stopPropagation());
                removeClickAway();
            }
        });

        return {
            isOpen,
            toggle,
            uiContextMenu,
            menu,
            menuStyle
        };
    }
});
</script>
