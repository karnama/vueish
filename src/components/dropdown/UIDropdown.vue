<template>
    <div ref="uiDropdown"
         aria-haspopup="true"
         class="ui-dropdown w-max relative">
        <slot name="trigger"
              :toggle="toggle"
              :show="show"
              :hide="hide" />

        <div v-if="isOpen"
             v-bind="$attrs"
             ref="dropdown"
             v-click-away="hide"
             class="dropdown absolute z-50 rounded overflow-scroll flex flex-col items-stretch shadow-lg
                    bg-white dark:bg-gray-600 ring-1 ring-opacity-5 ring-black dark:ring-white dark:ring-opacity-5"
             :style="dropdownStyle"
             role="group"
             @click.stop>
            <slot :toggle="toggle" :show="show" :hide="hide" />
        </div>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent, ref, reactive } from 'vue';
import type { PropType } from 'vue';
import clickAway from '@/directives/click-away';
import { getPxValue } from 'composables/style';

export default defineComponent({
    name: 'UIDropdown',

    directives: { clickAway },

    inheritAttrs: false,

    props: {
        /**
         * The horizontal position of the dropdown.
         */
        horizontal: {
            type: String as PropType<'right' | 'left'>,
            default: 'right',
            validator: (val: 'right' | 'left') => ['right', 'left'].includes(val)
        },

        /**
         * The vertical position of the dropdown.
         */
        vertical: {
            type: String as PropType<'top' | 'bottom'>,
            default: 'bottom',
            validator: (val: 'top' | 'bottom') => ['top', 'bottom'].includes(val)
        },

        /**
         * The max height of the dropdown with px or vh notation.
         */
        maxHeight: {
            type: String,
            default: '35vh'
        },

        /**
         * The width of the dropdown with px or vw notation if using atMousePosition
         * otherwise css rule accepted values.
         */
        // todo add cross validation when https://github.com/vuejs/vue-next/pull/3258 merged then default to auto
        width: {
            type: String,
            default: '250px'
        },

        /**
         * Whether the position should be where
         * the mouse is currently at or not.
         */
        atMousePosition: {
            type: Boolean,
            default: false
        }
    },

    emits: ['open', 'close'],

    expose: ['hide', 'show', 'toggle'],

    setup(props, ctx) {
        const isOpen = ref(false);
        const uiDropdown = ref<HTMLDivElement>();
        const dropdown = ref<HTMLDivElement>();
        const mousePos = reactive({
            x: 0,
            y: 0
        });

        const dropdownStyle = computed(() => {
            const style: Partial<CSSStyleDeclaration> = {};

            if (!uiDropdown.value || !isOpen.value) {
                return style;
            }

            const contentParameters: DOMRect = uiDropdown.value.getBoundingClientRect();
            const dropdownParameters = dropdown.value?.getBoundingClientRect();

            style.maxHeight = props.maxHeight;
            style.width = props.width;

            if (props.atMousePosition) {
                // convert viewport values to px
                const width = getPxValue(props.width);
                const maxHeight = getPxValue(props.width);

                // doesn't fit on the right
                if (document.documentElement.clientWidth - (contentParameters.x + mousePos.x) <= width) {
                    style.left = `${mousePos.x - width}px`;
                } else {
                    style.left = `${mousePos.x}px`;
                }

                // doesn't fit on the bottom
                if (document.documentElement.clientHeight - (contentParameters.y + mousePos.y) <= maxHeight) {
                    style.top = `${mousePos.y - (dropdownParameters ? dropdownParameters.height : maxHeight)}px`;
                } else {
                    style.top = `${mousePos.y}px`;
                }

                return style;
            }

            if (props.horizontal === 'left') {
                style.left = '0px';
            } else {
                style.right = '0px';
            }

            if (props.vertical === 'top') {
                style.bottom = `${contentParameters.height}px`;
            } else {
                style.top = `${contentParameters.height}px`;
            }

            return style;
        });

        const show = (event?: MouseEvent) => {
            if (event && props.atMousePosition) {
                mousePos.x = event.offsetX;
                mousePos.y = event.offsetY;
            } else {
                ctx.emit('open');
            }

            isOpen.value = true;
        };
        const hide = () => {
            isOpen.value = false;
            ctx.emit('close');
        };
        const toggle = (event?: MouseEvent) => {
            // If the dropdown is showing via contextmenu event, and the user has triggered
            // another contextmenu event then reshow the dropdown at the new mouse
            // position.
            const reshow = isOpen.value && props.atMousePosition && event?.type === 'contextmenu';

            if (isOpen.value) {
                hide();
            } else {
                show(event);
            }

            if (reshow) {
                show(event);
            }
        };

        return {
            isOpen,
            toggle,
            show,
            hide,
            uiDropdown,
            dropdown,
            dropdownStyle
        };
    }
});
</script>
