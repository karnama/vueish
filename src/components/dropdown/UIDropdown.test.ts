import { DOMWrapper, mount } from '@vue/test-utils';
import UIDropdown from './UIDropdown.vue';
import { h, nextTick } from 'vue';

describe('UIDropdown', () => {
    it('should correctly display', async () => {
        const wrapper = mount(UIDropdown, {
            slots: {
                default: h('div', { id: 'content' }),
                trigger: h('div', { id: 'trigger' })
            }
        });

        // @ts-expect-error - ts reckons this has the type never
        wrapper.vm.toggle();
        await nextTick();
        expect(document.body.lastElementChild).toMatchSnapshot();
        wrapper.unmount();
    });

    it('should display the given content when open', async () => {
        const wrapper = mount(UIDropdown, {
            slots: {
                default: h('div', { id: 'content' }),
                trigger: h('div', { id: 'trigger' })
            }
        });

        const documentWrapper = new DOMWrapper(document.body);

        expect(documentWrapper.find('#content').exists()).toBe(false);
        // @ts-expect-error - ts reckons this has the type never
        wrapper.vm.toggle();
        await nextTick();
        expect(documentWrapper.find('#content').exists()).toBe(true);
    });

    it('should emit the open and close event at appropriate times', () => {
        const wrapper = mount(UIDropdown, {
            slots: {
                default: h('div', { id: 'content' }),
                trigger: h('div', { id: 'trigger' })
            }
        });

        expect(wrapper.emitted()).not.toHaveProperty('open');
        expect(wrapper.emitted()).not.toHaveProperty('close');

        // @ts-expect-error - ts reckons this has the type never
        wrapper.vm.toggle();
        expect(wrapper.emitted('open')).toHaveLength(1);
        expect(wrapper.emitted()).not.toHaveProperty('close');

        // @ts-expect-error - ts reckons this has the type never
        wrapper.vm.toggle();
        expect(wrapper.emitted('close')).toHaveLength(1);
    });
});
