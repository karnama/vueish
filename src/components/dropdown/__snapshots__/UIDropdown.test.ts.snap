// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIDropdown should correctly display 1`] = `
<transition-stub>
  
  <div
    class="dropdown absolute z-50 rounded overflow-auto flex flex-col items-stretch shadow-lg bg-white dark:bg-gray-600 h-max ring-1 ring-opacity-5 ring-black dark:ring-white dark:ring-opacity-5"
    role="group"
    style="max-height: 35vh; width: 250px; left: 0px; top: 0px;"
  >
    
    <div
      id="content"
    />
    
  </div>
  
</transition-stub>
`;
